Add Support For Automatic Persisted Queries

pscarey opened the pull request:
Adds customisable support for Automatic Persisted Queries, with default settings compatible with `apollo-client`.

#110 already has discussion on this. We required support internally, so it's up to maintainers if support in this way is appropriate. Happy to make tweaks based on feedback if required.

Note this builds on #186.

mcollina commented:
I'm not convinced by this implementation. How do you plan to scale this horizontally? The queries will be stored only on one peer.

pscarey commented:
Interesting point - I think this is how Apollo Server implemented it, which is why I went down this road, then generalised the way the query is delivered.

The argument in favour would be that the cost of hashing a query the first time an instance sees it will still massively reduce the number of times the full query is sent over the wire.

---

Really, this entire interface or setting might be better off merged into the existing persisted queries solution, in a customisable manner, since there is no firm 'Best Practice', such that there is:

Persisted Query Settings:
-  `isPersistedQuery` check if the given GraphQL request/query is persisted or regular.
- `getHash` get the hash from a GraphQL persisted query request, or falsy if not found.
- `getQueryFromHash` return the query for a given hash, or falsy if not found.
- `getHashForQuery` return the hash for a given query, or falsy if not to be stored.
- `saveQuery` store the query for the given hash.
- `onlyPersisted` whether to allow persisted queries or not.

Example implementations would be:

Current Persisted Queries (Ahead of time):
```
{
  isPersistedQuery: (r) => r.persisted,
  getHash: (r) => r.query,
  getQueryFromHash: (hash) => persistedQueries[hash], // Load from memory
  // getHashForQuery and saveQuery are not required as we do not save new queries
}
```

Apollo Client Compatible Automatic Persisted Queries (Runtime):
```
{
  isPersistedQuery: (r) => r.extensions.persistedQuery.sha256Hash && r.extensions.persistedQuery.version === 1,
  getHash: (r) => r.extensions.persistedQuery.sha256Hash ,
  getQueryFromHash: (hash) => persistedQueries[hash], // Load from memory
  getHashForQuery: (query) => sha256(query),
  saveQuery: (hash, query) => persistedQueries[hash] = query // Store in memory
}
```

Cached Apollo Client Compatible Automatic Persisted Queries (Runtime + Shared Cache):
```
{
  isPersistedQuery: (r) => r.extensions.persistedQuery.sha256Hash && r.extensions.persistedQuery.version === 1,
  getHash: (r) => r.extensions.persistedQuery.sha256Hash ,
  getQueryFromHash: (hash) => redis.get(hash), // Load from redis
  getHashForQuery: (query) => sha256(query),
  saveQuery: (hash, query) => redis.set(hash, query) // Store in redis
}
```

mcollina commented:
My point is that the current interface (ahead of time) cannot be made dynamic without adding support for multiple instances of the server that store the query. Otherwise subsequent requests that hit a different server would not have the cached query.

In the code in this PR the “get” operation is synchronous and it should be made async (and likely moved in a common interface). In your list, the Persisted Query Settings:
-  `isPersistedQuery` check if the given GraphQL request/query is persisted or regular.
- `getHash` get the hash from a GraphQL persisted query request, or falsy if not found.
- `getQueryFromHash` return the query for a given hash, or falsy if not found. __must be async__
- `getHashForQuery` return the hash for a given query, or falsy if not to be stored.
- `saveQuery` store the query for the given hash. __must be async__
- `onlyPersisted` whether to allow persisted queries or not.

---

I’m also skeptical that adding a delay of a round trip to Redis is going to improve latency much in practice. We are trading the cost of sending a few hundred bytes from the client to fetching those from Redis. Given how HTTP/1.1 keepalive and TCP work, sending a few hundreds bytes more should have no impact.
The risk is to have an ahead-of-line blocking on the redis connection which will make thus caching moot.

Notably this makes a lot of sense if our query is massive.

pscarey commented:
I agree with your reasoning that it'd need to be async.

As a general rule, any persisted query requires that the payload reduction savings are greater than the lookup delay. How big the lookup delay is depends on the implementation.

My understanding is that:

Ahead of Time Persisted Queries (AoTPQ):
- Best performance in every use case.
- Hardest to set up, as all queries need to be compiled into hashes ahead of time.

Automated Persisted Queries (APQ):
- Using local memory will be slow for each of their first query, but then give equal performance compared to AoTPQ afterwards.
- The performance of this is basically equivalent to AoTPQ for long lived servers (i.e. not cloud functions).
- This is also the easiest to set up.

Cached APQ:
- Remove that first overhead at the cost of a roundtrip to the cache on every request.
- The performance for this would be worse than APQ for long lived servers, but better than APQ for short lived servers (cloud functions).
- The performance of this is heavily caveated on the payload reduction being sufficiently large to counteract the cost of the lookup time. However, it can reasonably be assumed than an in-datacenter Redis connection is much faster than e.g. a 3G mobile connection. Really this means it boils down to the use case, and the performance of the client.

---

Appreciate the responsiveness here, but I also want to flag that in order for us to migrate to fastify-gql, we need to support APQ as otherwise we'll lose backwards compatibility - obviously that doesn't mean that it's something fastify-gql should support - so if it's a no-go, we can fork.

mcollina commented:
Could you confirm that you are using "sticky sessions" for APQ? How can you guarantee that the same client goes to the same server otherwise? What i do not understand about APQ is if it's actually feasible in production. It seems very likely that a client talking to a different server might not get the cached one.

> I agree with your reasoning that it'd need to be async.

Then we are good, can you update the PR?

---

> Appreciate the responsiveness here, but I also want to flag that in order for us to migrate to fastify-gql, we need to support APQ as otherwise we'll lose backwards compatibility - obviously that doesn't mean that it's something fastify-gql should support - so if it's a no-go, we can fork.

I'm cautious about adding feature, I think this is legit and we should add it in one form or another.

pscarey commented:
I will do updates to both PR's tomorrow.

> Could you confirm that you are using "sticky sessions" for APQ? How can you guarantee that the same client goes to the same server otherwise? What i do not understand about APQ is if it's actually feasible in production. It seems very likely that a client talking to a different server might not get the cached one.

Our use case is as follows:

**Client:**
- Mobile App running `apollo-client` with approx 20 unique queries.
- Prior versions of the mobile app are supported, but we haven't tracked legacy queries (i.e. they are no longer in our master branch, but are still valid queries due to no breaking changes).
- `apollo-client` attempts to issue a query using the persisted queries hash. Expected API responses are as per [apollo-link-persisted-query](https://github.com/apollographql/apollo-link-persisted-queries#protocol).
- If the client receives the appropriate `PersistedQueryNotFound` response, then it resends the whole query. This causes the APQ cache to be populated for future requests.

**Server:**
- 3-4 replicas of our GraphQL API running in K8s.
- Queries are persisted when they are received in full, in local memory.
- After a short amount of usage, the most common queries will all have been persisted.
- After a little while longer, less frequently called queries are persisted.
- The machine continues its lifetime for a long period after, directly replying to persisted queries.

I hope this (very rough) diagram can covey the idea slightly better:
<img width="957" alt="Screen Shot 2020-06-18 at 10 01 22 pm" src="https://user-images.githubusercontent.com/20896758/85017699-46984980-b1af-11ea-9744-d2d116c5a91d.png">


Sticky sessions would ensure that the cache of the machine returning `PersistedQueryNotFound` is the one populated by the query, but are not strictly necessary, as the queries will still be persisted in a short period of time compared to the lifetime of the machine.

mcollina commented:
Thanks, it makes sense now. Please add a short version of that in the README.

pscarey commented:
I've reworked the API & tidied things up a lot based on the conversation above.

- The proposed API gives defaults which align with the existing options, support for Apollo's APQ, and room for additional customisation.
- Documentation and types are updated.
- All persisted query tests have been moved into a seperate file.

mcollina commented:
Can you rebase this on top of `master`?

pscarey commented:
Have merged the other changes in, diff should be clean.

Thanks.

pscarey commented on README.md:
```diff
@@ -92,51 +93,74 @@ app.get('/', async function (req, reply) {
+
+The API is the same as the `Prepared` default.
+
+#### ApolloAutomatic
```
Sure

pscarey commented on README.md:
```diff
@@ -648,8 +672,14 @@ __fastify-gql__ supports the following options:
     * `service.wsUrl`: The url of the websocket endpoint
     * `service.wsConnectionParams`: `Function` or `Object`
-* `persistedQueries`: A hash/query map to resolve the full query text using it's unique hash.
-* `onlyPersisted`: Boolean. Flag to control whether to allow graphql queries other than persisted. When `true`, it'll make the server reject any queries that are not present in the `persistedQueries` option above. It will also disable any ide available (playground/graphiql).
```
OK

pscarey commented on README.md:
````diff
@@ -92,51 +93,74 @@ app.get('/', async function (req, reply) {
+
+```js
+const persistedQuerySettings = {
+  ...GQL.PersistedQueryDefaults.ApolloAutomatic(),
````
Will add

pscarey commented on index.js:
```diff
@@ -32,6 +33,35 @@ const { ErrorWithProps, FEDERATED_ERROR } = require('./lib/errors')
+      notSupportedError: 'PersistedQueryNotSupported'
+    })
+  }
+}
```
Sure

pscarey commented on lib/routes.js:
```diff
@@ -145,29 +148,85 @@ module.exports = async function (app, opts) {
+    // Validate a query is present
     if (!query) {
-      return new BadRequest('Unknown query')
+      return new BadRequest('QueryNotFound')
```
No particular reason, have reverted.

pscarey commented:
I've:
- Updated the naming to Automatic, and clarified the imports in the README.
- Moved the `PersistedQueryDefaults` into a new file.
- Reverted the change of the naming of the query not found error back to `Unknown query`.
- Added support for `persistedQueries` and `onlyPersisted`, to automatically map to the `persistedQuerySettings` internally.
- Added tests for both using `persistedQuerySettings` and `persistedQueries`/`onlyPersisted`.

mcollina commented on README.md:
```diff
@@ -92,51 +93,101 @@ app.get('/', async function (req, reply) {
-### Disallowing unknown queries
+#### Automatic
+
+This default is compatible with `apollo-client`, and requires no additional tooling to set up at the cost of some performance. In order for this mode to be effective, you must have long lived server instances (i.e *not* cloud functions). This mode is also appropriate for public APIs where queries are not known ahead of time.
```
Adding some links to the relevant portion of apollo-client docs would be really handy, thanks.

pscarey replied:
Sure

mcollina commented on README.md:
````diff
@@ -92,51 +93,101 @@ app.get('/', async function (req, reply) {
 
+```js
+const persistedQuerySettings = {
+  ...GQL.PersistedQueryDefaults.Automatic(),
````
Can you add a require for GQL? Thanks

pscarey replied:
Sure

mcollina commented on lib/persistedQueryDefaults.js:
```diff
@@ -0,0 +1,32 @@
+const crypto = require('crypto')
+
+const PersistedQueryDefaults = {
+  Prepared: (persistedQueries) => ({
```
I prefer to keep naming of those lowercase

pscarey replied:
No worries

mcollina commented on lib/persistedQueryDefaults.js:
```diff
@@ -0,0 +1,32 @@
+const crypto = require('crypto')
```
`'use strict'` is missing at the top

pscarey replied:
Thanks - too used to TS

mcollina commented on README.md:
```diff
@@ -649,8 +700,16 @@ __fastify-gql__ supports the following options:
-* `onlyPersisted`: Boolean. Flag to control whether to allow graphql queries other than persisted. When `true`, it'll make the server reject any queries that are not present in the `persistedQueries` option above. It will also disable any ide available (playground/graphiql).
+* `persistedQueries`: A hash/query map to resolve the full query text using it's unique hash. Overrides `persistedQuerySettings`.
+* `onlyPersisted`: Boolean. Flag to control whether to allow graphql queries other than persisted. When `true`, it'll make the server reject any queries that are not present in the `persistedQueries` option above. It will also disable any ide available (playground/graphiql). Requires `persistedQueries` to be set, and overrides `persistedQuerySettings`.
+* `persistedQuerySettings`
```
Maybe `persistedQueryStorage` is a better name? those are functions.

pscarey replied:
Whilst most members are functions:
- `isPersistedQuery` and `getHash` functions extract data from the request payload. 
- `getQueryFromHash`, `saveQuery` are related to storage.
- `getHashForQuery` option is about hashing a query.
- `notFoundError` and `notSupportedError` are strings.

I think based on this breakdown, the options/settings are broader than "storage". Happy to change if you still would like, but maybe another name is better?

mcollina replied:
`persistedQueryProvider` maybe?

mcollina commented on lib/persistedQueryDefaults.js:
```diff
@@ -0,0 +1,32 @@
+  }),
+  Automatic: () => {
+    // Initialize only in the scope of this server instance
+    const AUTOMATIC_PERSISTED_QUERIES = {}
```
This must be a very large LRU cache instead, otherwise we are exposing ourselves to a DoS attack by memory exhaustion. check out https://www.npmjs.com/package/mnemonist.

pscarey replied:
This is a good point, I'll look into implementing it.

pscarey replied:
Have added

pscarey commented:
Minor fixes have been completed. Don't mind the long process - happy to get the prompt feedback, and I appreciate it's a non trivial set of changes.

Should other casing also be updated?
- `Prepared` -> `prepared`
- `PreparedOnly` -> `preparedOnly`
- `Automatic` -> `automatic`

mcollina commented on README.md:
```diff
@@ -92,51 +93,127 @@ app.get('/', async function (req, reply) {
-    add(x: Int, y: Int): Int
-  }
-`
+Alternatively the `peristedQueries` option may be used directly, which will be internally mapped to the `Prepared` default:
```
Prepared -> prepared.

mcollina commented on README.md:
```diff
@@ -649,8 +726,16 @@ __fastify-gql__ supports the following options:
+  * `getHashForQuery?: (query: string) => string`: Return the hash for a given query string. Do not provide if you want to skip saving new queries.
+  * `saveQuery?: (hash: string, query: string) => Promise<void>`: Save a query, given its hash.
+  * `notFoundError?: string`: An error message to return when `getQueryFromHash` returns no result. Defaults to `Bad Request`.
+  * `notSupportedError?: string`: An error message to return when a query matches `isPersistedQuery`, but returns no valid hash from `getHash`. Defaults to `Bad Request`.
```
Can you add `async` for the functions that should be async?

mcollina commented on index.d.ts:
```diff
@@ -157,6 +208,15 @@ declare namespace fastifyGQL {
+  export const persistedQueryDefaults: { 
+    Prepared: (persistedQueries: object) => PeristedQuerySettings
+    PreparedOnly: (persistedQueries: object) => PeristedQuerySettings
+    Automatic: (maxSize?: number) => PeristedQuerySettings
```
can you please use lower case here as well?

mcollina commented on lib/persistedQueryDefaults.js:
```diff
@@ -0,0 +1,36 @@
+      notFoundError: 'PersistedQueryNotFound',
+      notSupportedError: 'PersistedQueryNotSupported'
+    })
+  }
```
What is `r`? Can you please use descriptive variables names?

mcollina commented on lib/routes.js:
```diff
@@ -145,29 +148,85 @@ module.exports = async function (app, opts) {
+        const hash = getHash && getHash(body)
+        if (hash) {
+          // Load the query for the provided hash
+          query = getQueryFromHash && await getQueryFromHash(hash)
```
I think we should validate during plugin setup that if somebody is enabling persisted queries they should pass in a `getQueryFromHash` function.

mcollina commented on lib/routes.js:
```diff
@@ -145,29 +148,85 @@ module.exports = async function (app, opts) {
+    // Check for not yet persisted queries
+    if (enablePersistedQueries && !persisted) {
+      // If provided the getHashForQuery, saveQuery settings we save this query
+      const hash = getHashForQuery && getHashForQuery(query)
```
We have already computed the hash, so we should reuse what we computed before.

pscarey replied:
The process is broken down into two steps:
1. Retrieve the hash from the graphql request (i.e. accessing the appropriate property of the request object), and loading the appropriate query from the given hash.
2. If the query is not (yet) persisted (i.e. 1 has failed), compute the hash from the query string, and save it.

The part you've commented on is step 2, where we compute the hash. We don't have a computed hash at this stage, and we have been unable to find a valid hash in the query (as per the `!persisted` check).

mcollina commented:
@pscarey any updates on this? It would be great to land!

pscarey commented:
@mcollina Sorry about the delay - was away for a few days. Will be updating shortly.

pscarey commented:
All updated. FYI, we've been running this in production for over a week, and have had no issues.

pscarey commented on lib/routes.js:
```diff
@@ -145,29 +148,84 @@ module.exports = async function (app, opts) {
+
+  const enablePersistedQueries = !!persistedQueryProvider
+
   async function execute (body, request, reply) {
```
Have tidied this up - is substantially more readable, thanks.

pscarey commented:
Have implemented the feedback, but it looks like some changes from `master` are breaking CI?

mcollina commented:
Yes, we are fixing them today, unfortunately something broke in #147

mcollina commented:
I have the reverted the offending commit on master.

mcollina commented on lib/routes.js:
```diff
@@ -145,31 +148,95 @@ module.exports = async function (app, opts) {
+      const { query, operationName, variables } = body
+      return executeQuery(query, variables, operationName, request, reply)
+    }
+  }
```
I would do something like:

```js

function executeQueryWrap (body, request, reply) {
      const { query, operationName, variables } = body
      return executeQuery(query, variables, operationName, request, reply)
}

const execute = persistedQueryProvider ? executePersistedQuery : executeQueryWrap
```

pscarey replied:
Thanks, have updated
